meshmode/dof_array.py

```diff
@@ -448,34 +463,97 @@ def prg():
         for grp_start, (nel, ndof) in zip(group_starts, group_shapes)))
 
 
-def unflatten(actx: ArrayContext, discr,
-        ary: ArrayContainer,
-        ndofs_per_element_per_group: Optional[Iterable[int]] = None):
-    r"""Convert all 'flat' arrays returned by :func:`flatten` back to
+def _unflatten_group_sizes(discr, ndofs_per_element_per_group):
+    if ndofs_per_element_per_group is None:
+        ndofs_per_element_per_group = [
+                grp.nunit_dofs for grp in discr.groups]
+
+    group_shapes = [
+            (grp.nelements, ndofs_per_element)
+            for grp, ndofs_per_element
+            in zip(discr.groups, ndofs_per_element_per_group)]
+
+    group_sizes = [nel * ndof for nel, ndof in group_shapes]
+    group_starts = np.cumsum([0] + group_sizes)
+
+    return group_shapes, group_starts
+
+
+def unflatten(
+        actx: ArrayContext, discr, ary: ArrayContainer,
+        ndofs_per_element_per_group: Optional[Iterable[int]] = None, *,
+        strict: bool = True,
+        ) -> ArrayContainer:
+    r"""Convert all "flat" arrays returned by :func:`flatten` back to
     :class:`DOFArray`\ s.
 
+    This function recurses into the :class:`~arraycontext.ArrayContainer` for all
+    :class:`DOFArray`\ s. It assumes that all the :class:`DOFArray`\ s inside the
+    container *ary* have the same number of degrees of freedom per
+    element group in *discr*.
```

Comment:
Not sure I understand this sentence. Could you clarify?

Reply:
For example, if you give it an object array of `[element_centroids, nodes]`, where `element_centroids` has 1 DOF per element and `nodes` has `nunit_nodes` DOFs per element. Then it won't know how to unflatten the container as a whole, because it just gets one `ndofs_per_element_per_group` to customize the number of DOFs per element.

```suggestion
    :class:`DOFArray`\ s. All class:`DOFArray`\ s inside the
    container *ary* must have the same number of degrees of freedom,
    as dictated by `ndofs_per_element_per_group`.
```
?

Reply:
(Thread continued and resolved at https://github.com/inducer/meshmode/pull/196#discussion_r639932687.)